en/_posts/2024-05-23-understanding-github-artifact-attestations.md

```diff
@@ -0,0 +1,224 @@
+with some OID claims set.
+
+After we have an artifact and attestation, as a user, we need to be able to
+verify it before we use it. Verification works something like the following.
+The code for this is found in the [`gh attestations verify
+command`](https://github.com/cli/cli/tree/trunk/pkg/cmd/attestation/verify)
+for GitHub’s official CLI tool.
+
+1. The attestation is downloaded from the GitHub Attestations API using the
+   artifact’s digest.
+2. The attestation’s signature is verified against the public key used to sign
+   it, and to the certificate chain for either the GitHub Fulcio instance, or
+   the public Sigstore Fulcio instance (the root certificates are managed using
+   [TUF](https://theupdateframework.io/)).
+3. The expected values for the owner or repo given by the user are matched
+   against the signing certificate’s OID claims.
+
+Notice that nowhere here did we actually use the contents of the SLSA
+provenance for verification. We’ll discuss why this is below.
+
+## A Good User Experience
+
+By providing a GitHub Action, GitHub gives users the maximum amount of
+flexibility when integrating this into their GitHub Actions workflows. It’s
+really easy to add a job step to your workflow and pass it a path to your
+artifact file.
+
+```
+- name: Attest Build Provenance
+  uses: actions/attest-build-provenance@897ed5eab6ed058a474202017ada7f40bfa52940 # v1.0.0
+  with:
+    subject-path: "bin/my-artifact.tar.gz"
+```
+
+Easy-to-use UX is really important for security because it increases adoption.
+This can’t be understated and is an often overlooked aspect of security
+software.
+
+## Why SLSA Build L2?
+
+But if this is a fairly secure solution, why is this SLSA Build L2 and not L3?
+The main reason is that, by itself, the `attest-build-provenance` action
+doesn’t meet this requirement for [SLSA Build
+L3](https://slsa.dev/spec/v1.0/levels#build-l3-hardened-builds):
+
+prevent secret material used to sign the provenance from being accessible to the user-defined build steps.
+
+Because GitHub Actions are run in the same job VM with other build steps there
+is a chance that the user-defined build steps could access the secret material,
+namely the certificate’s private key provided by Fulcio, and use it for
+nefarious purposes. Normally an attacker, by compromising the build, might be
+able to use this key material to make up their own provenance for a malicious
+artifact and sign it with the key.
+
+However, as we’ll see, this kind of attack is mostly mitigated by using Sigstore.
+
+## SLSA Build L2+?
+
+By using Sigstore’s Fulcio the certificate used to sign the provenance contains
+much of the provenance information in its OID claims. These claims are signed
+by part of Fulcio instance’s certificate chain and thus cannot be modified by
+the user-defined build steps unless the Sigstore instance or GitHub’s OIDC
+provider are compromised.
+
+So while the SLSA provenance itself might be modified the certificate OID
```

Comment:
define "provenance". Here you mean intoto payload. The diagram will also help with that

Reply:
Yeah. I mean the SLSA provenance JSON (as opposed to the cert) which is why I wrote "SLSA provenance" rather than just "provenance". Maybe I could be more specific and say predicate, but I wanted to avoid needing to define too many terms.

Reply:
Updated to be clearer about use of provenance.